### ISSUE TITLE
Builds should run test/test-coverage, CLI/config option to switch this off

### ISSUE THREAD

**Issue by neomatrix369** [AdoptOpenJDK/openjdk-build#7]
Our Cloudbees/Jenkins jobs currently runs the native test suite (with the help of `jtreg`), with code coverage enabled (using `jcov`).

It would be great to have the new jobs be able to do the same.

Examples on how this is done can be see from the bash scripts attached the Jenkin jobs.

**Comment by tellison** [comment]
Yep, test is a significant area for further discussion and contribution.  We need to figure out what types of testing we need to do for different purposes (PR sanity check, build verification check, release level testing, etc).

Jtreg is fine as far as it goes, and should definitely feature in our overall test approach.

**Comment by neomatrix369** [comment]
A number of pull requests have been merged to help resolve this issue:
#22 
#24 
#31

**Comment by karianna** [comment]
@neomatrix369 What's left to go?  Can you summarise here?

**Comment by neomatrix369** [comment]
~Our Cloudbees/Jenkins jobs currently runs the native test suite (with the help of `jtreg`), with code coverage enabled (using `jcov`).~ (done)

~It would be great to have the new jobs be able to do the same. If necessary, provide CLI/config option to switch this off and just build the artifact.~ (done)

**If the test fails, the artifact isn't created *** <=== needs doing, although is easy to do, once the script is refactored a bit

~when it passes we have three tarballs, one with just the JDK binaries and the other two with the test coverage and test output reports included in the tarball, see https://adopt-openjdk.ci.cloudbees.com/view/OpenJDK/job/openjdk-1.8-linux-x86_64/~

**Sanity check of JCov reports - is it getting generate correctly, is it legible **
**Publishing link to the JCov reports from the workspace **

Items in bold are remaining.

**Comment by karianna** [comment]
@Bluevisitor reports that langtools jtreg fails:

```
/usr/bin/make -k -C  ../langtools/test  JT_JAVA=/home/jenkins/workspace/openjdk_build_x86-64_linux/openjdk/build/linux-x86_64-normal-server-release/images/j2sdk-image JTREG_HOME= TEST=jtreg jtreg
make[2]: Entering directory '/home/jenkins/workspace/openjdk_build_x86-64_linux/openjdk/langtools/test'
make[2]: *** No rule to make target '/bin/jtreg', needed by 'check-jtreg'.
```

**Comment by neomatrix369** [comment]
@karianna Known issue, I think he has made the smart decision of not enabling that test, its not on the priority list of the openjdk team to attend to it.

**Comment by karianna** [comment]
@neomatrix369 Why was it a 'default' then?  Or was that a 'default' that we set?

**Comment by neomatrix369** [comment]
@karianna dont follow which part was default? `/usr/bin/make -k -C  ../langtools/test...`  <=== this part of the command?

**Comment by neomatrix369** [comment]
Just had a look https://adopt-openjdk.ci.cloudbees.com/job/openjdk-1.8-linux-x86_64/592/consoleFull - last run (09-Apr-2017 14:17:19) on CloudBees, a snippet of langtools_test execution:
```
/usr/bin/gmake -k -C  ../langtools/test  JT_JAVA=/scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jdk8/build/linux-x86_64-normal-server-release/images/j2sdk-image JTREG_HOME=/scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jtreg TEST=jtreg jtreg
gmake[2]: Entering directory `/scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jdk8/langtools/test'
JT_JAVA=/scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jdk8/build/linux-x86_64-normal-server-release/images/j2sdk-image /scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jtreg/bin/jtreg \
  -J-Xmx512m \
  -vmoption:-Xmx768m \
  -a -ignore:quiet -v:fail,error,nopass \
          -r:/scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jdk8/langtools/build/linux-x86_64/test/langtools/jtreg/JTreport \
          -w:/scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jdk8/langtools/build/linux-x86_64/test/langtools/jtreg/JTwork \
          -jdk:/scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jdk8/build/linux-x86_64-normal-server-release/images/j2sdk-image \
           \
  -agentvm -concurrency:10 -timeoutFactor:5  \
          . \
|| ( status=$?; if [ $status -ge 3 ]; then exit $status ; fi ; \
    echo $status > /scratch/jenkins/workspace/openjdk-1.8-linux-x86_64/jdk8/langtools/build/linux-x86_64/test/langtools/jtreg/status.txt \
)
```
Might be behaving different due to difference in environments.

**Comment by neomatrix369** [comment]
`/usr/bin/make -k -C  ../langtools/test  JT_JAVA=/home/jenkins/workspace/openjdk_build_x86-64_linux/openjdk/build/linux-x86_64-normal-server-release/images/j2sdk-image JTREG_HOME= TEST=jtreg jtreg` - some how its forgetting the setting for `JTREG_HOME=....` for `langtools_...`

**Comment by karianna** [comment]
@neomatrix369 yeah the bit where langtools was part of the 'default' run

**Comment by neomatrix369** [comment]
The snippet @Bluevisitor pasted was from the log which showed one of the iterations, so far from the script I see only 2 components are iterated thru - `jdk_core` and `langtools`, `hotspot` used to me in the list but no run anymore.

That is another discussion, should we also run `hotspot` tests, they take a bitlong to finish.

**Comment by karianna** [comment]
The `hotspot` tests are actually the most vital in a way as they are platform dependant ones.  That said if we jenkins pipeline this, why not have that as a separate pipeline job...

**Comment by neomatrix369** [comment]
@karianna I'll open a new issue related to this one, since our test coverage might be lacking hotspot test coverage

**Comment by neomatrix369** [comment]
Its possible our github repo that clones the mercurial counterpart ( http://hg.openjdk.java.net/jdk8u/jdk8u
 ) isn't in sync with the source. The Makefiles in langtools are a bit older, can we pls check if our hg job on jenkins  is pull through all the changes regularly.

**Comment by karianna** [comment]
@neomatrix369 @gdams - my understanding is that the pull occurs every 30 minutes...

**Comment by neomatrix369** [comment]
@gdams i just checked git-hg (https://ci.adoptopenjdk.net/view/all/job/git-hg/158/console) its failing with these messages:
```
error: src refspec dev does not match any.
error: failed to push some refs to 'git@github.com:AdoptOpenJDK/openjdk-jdk8u.git'
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```
Something to worry about? The git repo seems to get the most recent changesets although not the last one:
```
2 weeks ago	asaha	Added tag jdk8u152-b02 for changeset d7045c4d4eab
```

**Comment by neomatrix369** [comment]
@karianna the langtools failure @Bluevisitor reported comes from these lines in the two build system files:

```
./openjdk/langtools/make/build.xml:    <target name="-check-jtreg.home" depends="-def-check">
./openjdk/langtools/make/build.xml:    <target name="-def-jtreg" unless="jtreg.defined" depends="-check-jtreg.home,-check-target.java.home">
./openjdk/langtools/test/Makefile:jtreg-tests: check-jtreg FRC
./openjdk/langtools/test/Makefile:check-jtreg: $(PRODUCT_HOME) $(JTREG)
./openjdk/langtools/test/Makefile:	jtreg javac javadoc javah javap jdeps jtreg-tests jtreg-summary check-jtreg \

```
Although these are not present in the latest `jdk8` clone (via `hg`)

**Comment by karianna** [comment]
This should get shifted to the test repo
